plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java

```diff
@@ -253,12 +250,34 @@ private Stream<DeltaLakeSplit> getSplits(
                 });
     }
 
-    private Set<Map<String, Optional<String>>> findPartitionsWithAtMostOneFile(List<AddFileEntry> addFileEntries)
+    private static Stream<AddFileEntry> filterValidDataFilesForOptimize(List<AddFileEntry> validDataFiles, long maxScannedFileSizeInBytes)
     {
-        return addFileEntries.stream().collect(groupingBy(AddFileEntry::getCanonicalPartitionValues, counting())).entrySet().stream()
-                .filter(entry -> entry.getValue() <= 1)
-                .map(Map.Entry::getKey)
-                .collect(toImmutableSet());
+        Map<Map<String, Optional<String>>, List<AddFileEntry>> queuedAddFileEntriesMap = new HashMap<>();
+        return validDataFiles.stream()
+                .flatMap(addFileEntry -> {
+                    Map<String, Optional<String>> canonicalPartitionValues = addFileEntry.getCanonicalPartitionValues();
+                    if (queuedAddFileEntriesMap.containsKey(canonicalPartitionValues)) {
+                        List<AddFileEntry> alreadyQueuedAddFileEntries = queuedAddFileEntriesMap.get(canonicalPartitionValues);
+                        if (alreadyQueuedAddFileEntries.isEmpty()) {
+                            return Stream.of(addFileEntry);
```

Comment:
it looks like empty list serves as a marker that the partition should be processed.
it's worth calling out with explicit comment

Reply:
is there a case where the List in the map has more than 1 element?
it doesn't seem so.

you can use `Optional<AddFileEntry>` for the pending entry

Reply:
Yes, indeed.

Optional would be a perfect holder.
Thanks for the idea.